### ISSUE TITLE
[Bug] Memory leak in fantasy models when the mean cache is not detached

### ISSUE THREAD

**Issue by saitcakmak** [cornellius-gp/gpytorch#2631]
# 🐛 Bug

This is a follow up to the memory leak reported in https://github.com/pytorch/botorch/issues/2728. Repeated operations on fantasized models within the `detach_test_caches(False)` context leads to indefinite increase in memory usage until the program crashes. Digging into it, I've isolated the leak to a single `detach()` call within the `DefaultPredictionStrategy._mean_cache`.

Things I've tried:
- Replacing `detach_test_caches(False)` with `detach_test_caches(True)` makes the leak go away.
- I haven't been able to reproduce with non-fantasy models. It could be that I haven't used large enough tensors to differentiate a leak from regular memory usage.
- Within `DefaultPredictionStrategy`, I've tried manually detaching each location `detach_test_caches` affects. Only the [`mean_cache`](https://github.com/cornellius-gp/gpytorch/blob/main/gpytorch/models/exact_prediction_strategies.py#L294) seems to affect the results. The other two locations didn't seem to make a meaningful difference in memory usage.
- I've tried not caching `_mean_cache` by removing the `@cached` decorator. The leak goes away if we remove caching. I then tried manually caching it as a simple attribute and that also produced the leak.

## To reproduce

** Code snippet to reproduce **
```python
import torch
from gpytorch import settings as gpt_settings
from gpytorch.distributions import MultivariateNormal
from gpytorch.kernels import RBFKernel
from gpytorch.likelihoods import GaussianLikelihood
from gpytorch.means import ConstantMean
from gpytorch.models import ExactGP

d = 10
mc_points = torch.rand(32, d, dtype=torch.double)


class SimpleGP(ExactGP):
    def __init__(self, train_inputs, train_targets):
        super().__init__(train_inputs, train_targets, GaussianLikelihood())
        self.mean_module = ConstantMean()
        self.covar_module = RBFKernel()

    def forward(self, x):
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        return MultivariateNormal(mean_x, covar_x)


gp = SimpleGP(
    train_inputs=torch.rand(256, d, dtype=torch.double),
    train_targets=torch.rand(256, dtype=torch.double),
).eval()
gp(torch.rand(5, d, dtype=torch.double))  # set the caches before fantasize.

X = torch.rand(128, 5, d, dtype=torch.double, requires_grad=False)
Y = torch.rand(128, 5, dtype=torch.double, requires_grad=False)

for i in range(1000):
    fantasy_model = gp.get_fantasy_model(inputs=X, targets=Y).eval()
    with gpt_settings.detach_test_caches(False):
        fantasy_model(mc_points)
    print(f"{i=}")
```

** Stack trace/error message **

Runs for a while until the process is killed.
```
i=0
i=1
i=2
i=3
i=4
...
i=294
i=295
i=296
i=297
i=298
zsh: killed     python frepro.py
```
Just before crashing, it was reporting ~100GB memory usage.

## Expected Behavior

Memory usage should not increase beyond a simple first step. The program should be able to run indefinitely.

## System information

**Please complete the following information:**
- GPyTorch 1.14
- Linear Operator 0.6
- Reproduces on both MacOS & CentOS

**Comment by Scienfitz** [comment]
+1
